src/app/icd/client/ICDRefreshKeyInfo.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ICDRefreshKeyInfo.h"
+#include "CheckInDelegate.h"
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandPathParams.h>
+#include <app/InteractionModelEngine.h>
+#include <app/InteractionModelTimeout.h>
+#include <app/OperationalSessionSetup.h>
+
+namespace chip {
+namespace app {
+
+ICDRefreshKeyInfo::ICDRefreshKeyInfo(CheckInDelegate * apCheckInDelegate, const ICDClientInfo aICDClientInfo,
+                                     const ICDClientStorage * aICDClientStorage) :
+    mOnConnectedCallback(HandleDeviceConnected, this),
+    mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+{
+    mpCheckInDelegate  = apCheckInDelegate;
+    mICDClientInfo     = aICDClientInfo;
+    mpICDClientStorage = const_cast<ICDClientStorage *>(aICDClientStorage);
+}
+
+void ICDRefreshKeyInfo::OnResponse(chip::app::CommandSender * apCommandSender, const chip::app::ConcreteCommandPath & aPath,
+                                   const chip::app::StatusIB & aStatus, chip::TLV::TLVReader * aData)
+{
+    app::Clusters::IcdManagement::Commands::RegisterClientResponse::DecodableType responseData;
+    mError = DataModel::Decode(*aData, responseData);
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to decode re-registration response data: %" CHIP_ERROR_FORMAT, mError.Format());
+        return;
+    }
+    // Update the ICDClientInfo with new key and start counter and store it to persistence
+    mICDClientInfo.start_icd_counter = responseData.ICDCounter;
+    mError                           = mpICDClientStorage->SetKey(mICDClientInfo, mNewKey.Span());
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to store the new key after re-registration: %" CHIP_ERROR_FORMAT, mError.Format());
+    }
+}
+void ICDRefreshKeyInfo::OnError(const chip::app::CommandSender * apCommandSender, CHIP_ERROR aError)
+{
+    mError = aError;
+}
+
+void ICDRefreshKeyInfo::OnDone(chip::app::CommandSender * apCommandSender)
+{
+    if (mError == CHIP_NO_ERROR)
+    {
+        mpCheckInDelegate->OnRegistrationUpdateComplete(mICDClientInfo, this);
+        return;
+    }
+    mpCheckInDelegate->OnRegistrationUpdateFailure(this);
```

Comment:
This should probably pass in the error, so that the callee can log or take other appropriate action as desired.

Reply:
Done.